bee-vote/src/fpc.rs

```diff
@@ -0,0 +1,587 @@
+        Ok(Arc::try_unwrap(all_queried_opinions).unwrap().into_inner())
+    }
+
+    /// Run a query on a given `OpinionGiver`, to generate opinions on the voting object.
+    async fn do_query(
+        query_ids: &QueryObjects,
+        vote_map: Arc<RwLock<HashMap<VoteObject, Opinions>>>,
+        all_queried_opinions: Arc<RwLock<Vec<QueriedOpinions>>>,
+        opinion_giver: &mut Box<dyn OpinionGiver>,
+        selected_count: u32,
+    ) {
+        // Get opinions from the `OpinionGiver`.
+        let opinions = if let Ok(opinions) = opinion_giver.query(query_ids) {
+            // Sanity check.
+            if opinions.len() != query_ids.conflict_objects.len() + query_ids.timestamp_objects.len() {
+                return;
+            } else {
+                opinions
+            }
+        } else {
+            return;
+        };
+
+        let mut queried_opinions = QueriedOpinions {
+            opinion_giver_id: opinion_giver.id().to_string(),
+            opinions: HashMap::new(),
+            times_counted: selected_count,
+        };
+
+        let mut vote_map_guard = vote_map.write().await;
+
+        // Get opinions on a voting object and add to vote map.
+        let mut query_voting_objects = |objects: &Vec<VoteObject>| {
+            for (i, object) in objects.iter().enumerate() {
+                let mut votes = vote_map_guard
+                    .get(object)
+                    .map_or(Opinions::new(vec![]), |opt| opt.clone());
+
+                for _ in 0..selected_count {
+                    votes.push(opinions[i]);
+                }
+
+                queried_opinions.opinions.insert(*object, opinions[i]);
+
+                if vote_map_guard.contains_key(object) {
+                    *vote_map_guard.get_mut(object).unwrap() = votes;
+                } else {
+                    vote_map_guard.insert(*object, votes);
+                }
+            }
+        };
+
+        query_voting_objects(&query_ids.conflict_objects);
+        query_voting_objects(&query_ids.timestamp_objects);
+
+        all_queried_opinions.write().await.push(queried_opinions);
+    }
+
+    /// Get the IDs of all `VoteContext`s currently in the voting pool.
+    async fn vote_context_ids(&self) -> QueryObjects {
+        let context_guard = self.contexts.read().await;
+        let mut conflict_objects = vec![];
+        let mut timestamp_objects = vec![];
+
+        for (_, context) in context_guard.iter() {
+            match context.object() {
+                VoteObject::Conflict(_) => {
+                    conflict_objects.push(context.object());
+                }
+                VoteObject::Timestamp(_) => {
+                    timestamp_objects.push(context.object());
+                }
+            }
+        }
+
+        QueryObjects {
+            conflict_objects,
+            timestamp_objects,
+        }
+    }
+
+    fn round_thresholds(&self, ctx: &VoteContext) -> (f64, f64) {
```

Comment:
wdyt about renaming to `round_opinion_change_thresholds()`?

Could you comment this? To make sure I understand:

/// Return the uniform range used to select the threshold at which a node will change it's opinion based on votes.

Reply:
I think maybe something like `calculate_round_thresholds()`? The function doesn't change anything (and takes a shared reference to `self`), but instead uses the context to return the correct set of thresholds for the current round.